src/allocator_defrag.c

```diff
@@ -0,0 +1,478 @@
+
+
+/* -----------------------------------------------------------------------------
+ * Alloc/Free API that are cooperative with defrag
+ * -------------------------------------------------------------------------- */
+/* Allocation and free functions that bypass the thread cache
+ * and go straight to the allocator arena bins.
+ * Currently implemented only for jemalloc. Used for online defragmentation. */
+void *defrag_jemalloc_alloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void defrag_jemalloc_free(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
+#define SIZE_CLASS_GROUP_SZ 4
+
+#define LG_QUANTOM_OFFSET_3 ((64 >> LG_QUANTOM_8_FIRST_POW2) - 1)
+#define LG_QUANTOM_OFFSET_4 (64 >> 4)
+
+#define get_binind_normal(_sz, _offset, _last_sz_pow2)                                                                 \
+    ((SIZE_CLASS_GROUP_SZ - (((1 << (_last_sz_pow2)) - (_sz)) >> ((_last_sz_pow2) - LG_QUANTOM_8_FIRST_POW2))) +       \
+     (((_last_sz_pow2) - (LG_QUANTOM_8_FIRST_POW2 + 3)) - 1) * SIZE_CLASS_GROUP_SZ + (_offset))
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions*/
+inline unsigned jemalloc_sz2binind_lgq3(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 24, 32, 40, 48, 56, 64
+        return (sz >> 3) - 1;
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_3, last_sz_in_group_pow2);
+}
+
+inline unsigned jemalloc_sz2binind_lgq4(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 32, 48, 64
+        return (sz >> 4);
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_4, last_sz_in_group_pow2);
+}
+
+/* -----------------------------------------------------------------------------
+ * Get INFO string about the defrag.
+ * -------------------------------------------------------------------------- */
+/*
+ * add defrag info string into info
+ */
+sds defrag_jemalloc_get_fragmentation_info(sds info) {
```

Comment:
```suggestion
sds defrag_cat_fragmentation_info(sds info) {
```
It's not clear that this function invalidates the input based on the current naming with a concat, it doesn't just return the info.